hazelcast/src/main/java/com/hazelcast/internal/util/ConcurrencyUtil.java

```diff
@@ -47,7 +47,7 @@ public String toString() {
 
     // Default executor for async callbacks: ForkJoinPool.commonPool() or a thread-per-task executor when
     // the common pool does not support parallelism
-    public static final Executor DEFAULT_ASYNC_EXECUTOR;
+    private static Executor defaultAsyncExecutor;
```

Comment:
Since it's not `final` it should be `volatile`. This field will be accessed on dependent stage registration/execution, so it will be relatively frequently accessed by async API users.

The alternative would be to keep the field as `final` and revert to your previous solution using reflection and limit its usage when tests run on JDK < 12 ([since which](https://bugs.openjdk.java.net/browse/JDK-8217225) it is not possible to access `Field#modifiers` field via reflection). So PR builder, compatibility, nightly, sonar etc builds will benefit from proper test names in `FJP#commonPool` threads, except build jobs executing tests on newer JDKs.

wdyt? /cc @jerrinot

Reply:
> Since it's not final it should be volatile.
Why ? I mean if we are sure that this is set before any thread starts, do we still need `final` /`volatile` ? 
My knowledge of JMM says it should be ok but since JMM is a giant beast, I am asking to be sure.

Another question, if this is not safe. How we are ok with removing the modifier via reflection in the first place ?
Then, this should cause random failures on our test suite which is not something we want.

Reply:
> I mean if we are sure that this is set before any thread starts, do we still need final /volatile ?

You are right about static initializers. I read more and [JLS 12.4.2](https://docs.oracle.com/javase/specs/jls/se8/html/jls-12.html#jls-12.4.2) indicates that class initialization is executed under a lock (or at least with memory effects of a lock acquire-release, but let’s not get lost in the details) and the same happens for initialized class state reading -> safe publication is ensured.

> Another question, if this is not safe. How we are ok with removing the modifier via reflection in the first place ?
Then, this should cause random failures on our test suite which is not something we want.

Great observation. I did a bit of digging and here’s what I found:
when the static field is not `volatile`, reflective `Field#set` uses [Unsafe#putObject](https://github.com/openjdk/jdk/blob/jdk8-b120/jdk/src/share/classes/sun/reflect/UnsafeStaticObjectFieldAccessorImpl.java#L82). When the field is `volatile`, then it uses [`Unsafe#putObjectVolatile`](https://github.com/openjdk/jdk/blob/jdk8-b120/jdk/src/share/classes/sun/reflect/UnsafeQualifiedStaticObjectFieldAccessorImpl.java#L84) which is “identical to putObject with volatile store semantics”.
So when the field is `final`, since it is not `volatile` and we remove the `final` modifier, reflective `Field#set` will use plain memory order. It seems that in order to ensure `volatile` memory semantics, we should directly use `Unsafe` ourselves:
```
Field field = ConcurrencyUtil.class.getDeclaredField("DEFAULT_ASYNC_EXECUTOR");
Object base = UnsafeUtil.UNSAFE.staticFieldBase(field);
long fieldOffset = UnsafeUtil.UNSAFE.staticFieldOffset(field);
UnsafeUtil.UNSAFE.putObjectVolatile(base, fieldOffset, "UPDATED");
```

(not saying this is a good idea though...)

Reply:
Actually my reasoning is much simpler and it is not related to static initializers. 

We have a static non-volatile field somewhere in a class. 

static int x; 

The program starts and sets this field when there is only one(main) thread. 
x = 5;

Any other thread starts after this. From this point if we set this field to any value we don't have visibility guarantee. 
But if we are sure that no-one touches it, then we are sure that every thread will see value 5. 

So there is no need for a lock of static initializers anyway for this to work correctly. 


The following two guarantees that (from https://docs.oracle.com/javase/6/docs/api/java/util/concurrent/package-summary.html#MemoryVisibility)
-> Each action in a thread happens-before every action in that thread that comes later in the program's order.
-> A call to start on a thread happens-before any action in the started thread.